include/cuda/std/array

```diff
@@ -0,0 +1,35 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the CUDA Toolkit, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_ARRAY
+#define _CUDA_ARRAY
+
+#include "cassert"
+#include "cstdint"
+#include "limits"
+#include "type_traits"
+#include "iterator"
+#include "utility"
+
+#include <initializer_list>
```

Comment:
@jrhemstad This is mostly what I was talking about. We can't use `cuda::std::initializer_list`, due to the implicit construction of a `std::initializer_list`.

The compiler might insert something like the below, and if we are really certain on using a `cuda::std::initializer_list` we would need to add an implicit conversion which makes 'our' `initializer_list` behave weird: https://gcc.godbolt.org/z/EG739eEdP

http://eel.is/c++draft/dcl.init.list#5

Reply:
Yeah, I don't see how we're going to get around this. `intializer_list` is a type that has special rules that the compiler follows that we can't really emulate with `cuda::std::initializer_list`.